src/jit/optimizer.cpp

```diff
@@ -1416,410 +1418,1018 @@ void Compiler::optCheckPreds()
 
 #endif // DEBUG
 
-/*****************************************************************************
- * Find the natural loops, using dominators. Note that the test for
- * a loop is slightly different from the standard one, because we have
- * not done a depth first reordering of the basic blocks.
- */
-
-void Compiler::optFindNaturalLoops()
+namespace
 {
-#ifdef DEBUG
-    if (verbose)
+//------------------------------------------------------------------------
+// LoopSearch: Class that handles scanning a range of blocks to detect a loop,
+//             moving blocks to make the loop body contiguous, and recording
+//             the loop.
+//
+// We will use the following terminology:
+//   HEAD    - the basic block that flows into the loop ENTRY block (Currently MUST be lexically before entry).
+//             Not part of the looping of the loop.
+//   FIRST   - the lexically first basic block (in bbNext order) within this loop.  (May be part of a nested loop,
+//             but not the outer loop. ???)
+//   TOP     - the target of the backward edge from BOTTOM. In most cases FIRST and TOP are the same.
```

Comment:
>In most cases FIRST and TOP are the same. [](start = 60, length = 41)

"In most cases FIRST and TOP are the same." It would be useful if the comment gave an example of when they are NOT the same.